From 14d4760aacb8896f99422c06d100e5231e09e797 Mon Sep 17 00:00:00 2001
From: Sayan Chowdhury <schowdhury@microsoft.com>
Date: Thu, 24 Aug 2023 00:00:00 +0530
Subject: Add verity hash passthrough

Read the verity hash from the kernel binary and pass it to the running
system via the kernel command line

The patch is prepared using the coreos/grub PRs, picking the only
required ones, and dropping the others. The README.md file in this
directory contains more contexual information along with the commits
used to create the patch.

Authored-by: Matthew Garrett <mjg59@coreos.com>
Signed-off-by: Sayan Chowdhury <schowdhury@microsoft.com>
---

diff -Naur a/grub-core/loader/efi/linux.c b/grub-core/loader/efi/linux.c
--- a/grub-core/loader/efi/linux.c	2024-09-05 16:04:35.466205483 -0000
+++ b/grub-core/loader/efi/linux.c	2024-09-05 16:54:36.760503901 -0000
@@ -38,6 +38,8 @@
 #include <grub/lib/cmdline.h>
 #include <grub/verify.h>
 
+#include <grub/verity-hash.h>
+
 GRUB_MOD_LICENSE ("GPLv3+");
 
 static grub_dl_t my_mod;
@@ -833,7 +835,8 @@
   kernel = NULL;
 #endif
 
-  cmdline_size = grub_loader_cmdline_size (argc, argv) + sizeof (LINUX_IMAGE);
+  cmdline_size = grub_loader_cmdline_size (argc, argv) + sizeof (LINUX_IMAGE)
+	+ VERITY_CMDLINE_LENGTH;
   linux_args = grub_malloc (cmdline_size);
   if (!linux_args)
     {
@@ -850,6 +853,7 @@
 
   if (grub_errno == GRUB_ERR_NONE)
     {
+      grub_pass_verity_hash (kernel_addr, linux_args, cmdline_size);
       grub_loader_set (grub_linux_boot, grub_linux_unload, 0);
       loaded = 1;
     }
diff -Naur a/grub-core/loader/i386/efi/linux.c b/grub-core/loader/i386/efi/linux.c
--- a/grub-core/loader/i386/efi/linux.c	2024-09-11 15:08:38.463525457 -0000
+++ b/grub-core/loader/i386/efi/linux.c	2024-09-11 15:10:30.732085424 -0000
@@ -31,6 +31,8 @@
 #include <grub/tpm.h>
 #include <grub/safemath.h>
 
+#include <grub/verity-hash.h>
+
 GRUB_MOD_LICENSE ("GPLv3+");
 
 static grub_dl_t my_mod;
@@ -466,7 +468,7 @@
   grub_dprintf ("linux", "new lh is at %p\n", lh);
 
   grub_dprintf ("linux", "setting up cmdline\n");
-  cmdline = kernel_alloc (KERNEL_MEM, lh->cmdline_size + 1,
+  cmdline = kernel_alloc (KERNEL_MEM, lh->cmdline_size + 1 + VERITY_CMDLINE_LENGTH,
 			  GRUB_EFI_LOADER_DATA,
 			  N_("can't allocate cmdline"));
   if (!cmdline)
@@ -479,6 +481,9 @@
 			      lh->cmdline_size - (sizeof (LINUX_IMAGE) - 1),
 			      GRUB_VERIFY_KERNEL_CMDLINE);
 
+  grub_pass_verity_hash (kernel, cmdline + sizeof (LINUX_IMAGE) - 1,
+			 lh->cmdline_size - (sizeof (LINUX_IMAGE) - 1) + VERITY_CMDLINE_LENGTH);
+
   grub_dprintf ("linux", "cmdline:%s\n", cmdline);
   grub_dprintf ("linux", "setting lh->cmd_line_ptr to 0x%08x\n",
 		LOW_U32(cmdline));
diff -Naur a/grub-core/loader/i386/pc/linux.c b/grub-core/loader/i386/pc/linux.c
--- a/grub-core/loader/i386/pc/linux.c	2024-09-11 15:08:38.464525462 -0000
+++ b/grub-core/loader/i386/pc/linux.c	2024-09-11 15:12:26.552628521 -0000
@@ -38,6 +38,8 @@
 #include <grub/safemath.h>
 #include <grub/efi/sb.h>
 
+#include <grub/verity-hash.h>
+
 GRUB_MOD_LICENSE ("GPLv3+");
 
 #define GRUB_LINUX_CL_OFFSET		0x9000
@@ -358,6 +360,11 @@
   if (err)
     goto fail;
 
+  grub_pass_verity_hash (&lh,
+			 (char *)grub_linux_real_chunk
+			 + GRUB_LINUX_CL_OFFSET + sizeof (LINUX_IMAGE) - 1,
+			 maximal_cmdline_size);
+
   if (grub_linux_is_bzimage)
     grub_linux_prot_target = GRUB_LINUX_BZIMAGE_ADDR;
   else
diff -Naur a/include/grub/verity-hash.h b/include/grub/verity-hash.h
--- a/include/grub/verity-hash.h	1970-01-01 00:00:00.000000000 -0000
+++ b/include/grub/verity-hash.h	2024-09-11 15:07:54.473295324 -0000
@@ -0,0 +1,51 @@
+/* CoreOS verity hash */
+
+#define VERITY_ARG " verity.usrhash="
+#define VERITY_ARG_LENGTH (sizeof (VERITY_ARG) - 1)
+#define VERITY_HASH_LENGTH 64
+#define VERITY_CMDLINE_LENGTH ((VERITY_ARG_LENGTH)+(VERITY_HASH_LENGTH))
+
+#if defined(__aarch64__)
+# define VERITY_HASH_OFFSET 512
+#elif defined(__i386__) || defined(__amd64__)
+# define VERITY_HASH_OFFSET 0x40
+#else
+# error Unsupported arch
+#endif
+
+
+/**
+ * grub_pass_verity_hash - Reads the CoreOS verity hash value from a well known
+ * kernel image offset and adds a kernel command line argument for it.
+ *
+ * @pImage: Kernel image buffer.
+ * @cmdline: Kernel command line buffer.
+ * @cmdline_max_len: Kernel command line buffer length.
+ */
+
+static inline void grub_pass_verity_hash(const void *pImage,
+					 char *cmdline,
+					 grub_size_t cmdline_max_len)
+{
+  const char *buf = pImage;
+  grub_size_t cmdline_len;
+  int i;
+
+  for (i=VERITY_HASH_OFFSET; i<VERITY_HASH_OFFSET + VERITY_HASH_LENGTH; i++)
+    {
+      if (buf[i] < '0' || buf[i] > '9') // Not a number
+	if (buf[i] < 'a' || buf[i] > 'f') // Not a hex letter
+	  return;
+    }
+
+  cmdline_len = grub_strlen(cmdline);
+  if (cmdline_len + VERITY_CMDLINE_LENGTH > cmdline_max_len)
+    return;
+
+  grub_memcpy (cmdline + cmdline_len, VERITY_ARG, VERITY_ARG_LENGTH);
+  cmdline_len += VERITY_ARG_LENGTH;
+  grub_memcpy (cmdline + cmdline_len, buf + VERITY_HASH_OFFSET,
+	       VERITY_HASH_LENGTH);
+  cmdline_len += VERITY_HASH_LENGTH;
+  cmdline[cmdline_len] = '\0';
+}
